libr/anal/block.c

```diff
@@ -22,74 +35,95 @@ R_API RAnalBlock *r_anal_block_split(RAnalBlock *bb, ut64 addr) {
 	return NULL;
 }
 
-R_API RAnalBlock *r_anal_block_new(RAnal *a, ut64 addr, int size) {
+R_API RAnalBlock *r_anal_block_new(RAnal *a, ut64 addr, ut64 size) {
 	RAnalBlock *b = r_anal_bb_new ();
-	if (b) {
-		b->addr = addr;
-		b->size = size;
+	if (!b) {
+		return NULL;
 	}
+	b->addr = addr;
+	b->size = size;
 	b->anal = a;
 	return b;
 }
 
-R_API void r_anal_block_free(RAnalBlock *bb) {
+static void __block_free(RAnalBlock *bb) {
 	r_anal_bb_free (bb);
 }
 
-static ut64 __bbHashKey(ut64 addr) {
-	return addr >> 4;
+void __block_free_rb(RBNode *node, void *user) {
+	RAnalBlock *block = container_of (node, RAnalBlock, rb);
+	__block_free (block);
 }
 
-R_API RAnalBlock *r_anal_get_block(RAnal *anal, ut64 addr) {
-	BBAPI_PRELUDE (NULL);
-	// XXX use the rbtree api
-	const ut64 k = __bbHashKey (addr);
-	RList *list = ht_up_find (anal->ht_bbs, k, NULL);
-	if (list) {
-		RAnalBlock *b;
-		RListIter *iter;
-		r_list_foreach (list, iter, b) {
-			if (addr >= b->addr && addr < (b->addr + b->size)) {
-				return b;
-			}
-		}
+R_API RAnalBlock *r_anal_get_block_at(RAnal *anal, ut64 addr) {
+	// TODO: this might be a bit faster using a ht
+	RBNode *node = r_rbtree_find (anal->bb_tree, &addr, __bb_addr_cmp, NULL);
+	if (!node) {
+		return NULL;
+	}
+	return container_of (node, RAnalBlock, rb);
+}
+
+R_API RAnalBlock *r_anal_get_block_in(RAnal *anal, ut64 addr) {
+	BBAPI_PRELUDE(x)
+	RBNode *node = r_rbtree_lower_bound (anal->bb_tree, &addr, __bb_addr_cmp, NULL);
+	if (!node) {
+		return NULL;
+	}
+	RAnalBlock *block = container_of (node, RAnalBlock, rb);
+	if (addr - block->addr < block->size) {
+		return block;
 	}
 	return NULL;
 }
 
+R_API void r_anal_get_blocks_intersect(RAnal *anal, ut64 addr, ut64 size, R_OUT RPVector *out) {
```

Comment:
wha about returning RList* instead of using the R_OUT parameter? im aware about the performance reasons..

Reply:
fine for me